tests/unit/test_endpoint_provider.py

```diff
@@ -544,3 +553,141 @@ def test_deeply_nested(self):
+        metadata_yaml = textwrap.dedent(
+            """
+            provides:
+              metrics-endpoint:
+                interface: prometheus_scrape
+            """
+        )
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args)
+
+            self.sandbox = TempFolderSandbox()
+            alert_rules_path = os.path.join(self.sandbox.root, "alert_rules")
+            for filepath, contents in rules.items():
+                self.sandbox.put_file(os.path.join(alert_rules_path, filepath), contents)
+
+            self.provider = MetricsEndpointProvider(
+                self, jobs=JOBS, alert_rules_path=alert_rules_path
+            )
+
+    return ProviderCharm
+
+
+class TestAlertRulesDoNotContainUnitNumber(unittest.TestCase):
+    """Feature: Consumer lib must not add a unit number to rule filters or labels.
+
+    Background: In lieu of SDI or pydantic, schema verification is done in unit tests.
+    """
+
+    def setUp(self):
+        pass
+
+    @patch_network_get(private_address="192.0.8.2")
+    def test_alert_rules_with_topology(self):
+        """Scenario: Since only the leader sets rel data, unit number can't be used as a filter."""
+        # GIVEN a provider charm has rule file(s) with a topology stub in the expr
+        provider_charm = provider_with_rules_factory(
+            {
+                "with_topology.rule": yaml.safe_dump(
+                    {"alert": "with_topology", "expr": "up{%%juju_topology%%} == 0"}
+                )
+            }
+        )
+
+        # WHEN relation data is generated
+        self.harness = Harness(provider_charm, meta=provider_charm.metadata_yaml)
+        self.harness.set_leader(True)
+        rel_id = self.harness.add_relation("metrics-endpoint", "prom")
+        self.harness.add_relation_unit(rel_id, "prom/0")
+        self.harness.begin_with_initial_hooks()
+
+        # THEN relation data does not contain juju_unit in expr or labels
+        relation = self.harness.charm.model.get_relation("metrics-endpoint")
+        alert_rules = json.loads(relation.data[self.harness.charm.app].get("alert_rules"))
+        for group in alert_rules["groups"]:
+            with self.subTest(group=group["name"]):
+                for rule in group["rules"]:
+                    with self.subTest(rule=rule["alert"]):
+                        self.assertNotIn("juju_unit", rule["labels"])
+                        self.assertNotIn("juju_unit=", rule["expr"])
+
+                        # These two are to prevent false negatives in case something big changes
+                        self.assertIn("juju_application", rule["labels"])
+                        self.assertIn("juju_application=", rule["expr"])
+
+        self.harness.cleanup()
+
+    @patch_network_get(private_address="192.0.8.2")
+    def test_alert_rules_without_topology(self):
+        """Scenario: Free standing rules shouldn't have any topology whatsoever, including unit."""
```

Comment:
This is news to me : "Free standing rules shouldn't have any topology whatsoever" ?  Isn't Juju Topology injected into **all** alert rules expressions and labels ?

Reply:
Correct, this is intended for "generic" rules that are not associated with any charm in particular.

Reply:
I do not see why the `MetricsEndpointProvider` is concerned with "free standing rules". Provider has the same behaviour regardless of where it is used isnt' it. In all cases it adds topology labels into alert rules (without the unit after this commit). Isn't this correct ? If so this comment not necessary. More importantly as mentioned above this test is redundant.